cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -0,0 +1,419 @@
+ */
+public class DataprocRuntimeJobManager implements RuntimeJobManager {
+  private static final Logger LOG = LoggerFactory.getLogger(DataprocRuntimeJobManager.class);
+
+  private static final String DATAPROC_GOOGLEAPIS = "-dataproc.googleapis.com:443";
+  private static final String CDAP_GCS_ROOT = "cdap-job";
+  // dataproc job properties
+  private static final String CDAP_RUNTIME_NAMESPACE = "cdap.runtime.namespace";
+  private static final String CDAP_RUNTIME_APPLICATION = "cdap.runtime.application";
+  private static final String CDAP_RUNTIME_PROGRAM = "cdap.runtime.program";
+  private static final String CDAP_RUNTIME_PROGRAM_TYPE = "cdap.runtime.program.type";
+  private static final String CDAP_RUNTIME_RUNID = "cdap.runtime.runid";
+
+  private final String clusterName;
+  private final GoogleCredentials credentials;
+  private final String projectId;
+  private final String region;
+  private final String bucket;
+  private final Map<String, String> labels;
+
+  private Storage storageClient;
+  private JobControllerClient jobControllerClient;
+
+  /**
+   * Created by dataproc provisioner with properties that are needed by dataproc runtime job manager.
+   *
+   * @param clusterName name of the cluster on which job should be submitted
+   * @param credentials google credentials
+   * @param projectId project id
+   * @param region region
+   * @param bucket gcs bucket
+   * @param labels system labels to be added on dataproc job
+   */
+  public DataprocRuntimeJobManager(String clusterName, GoogleCredentials credentials, String projectId, String region,
+                                   String bucket, Map<String, String> labels) {
+    this.clusterName = clusterName;
+    this.credentials = credentials;
+    this.projectId = projectId;
+    this.region = region;
+    this.bucket = bucket;
+    this.labels = Collections.unmodifiableMap(new HashMap<>(labels));
+  }
+
+  @Override
+  public void initialize() throws Exception {
+    // instantiate a gcs client
+    this.storageClient = StorageOptions.newBuilder().setProjectId(projectId)
+      .setCredentials(credentials).build().getService();
+
+    // instantiate a dataproc job controller client
+    CredentialsProvider credentialsProvider = FixedCredentialsProvider.create(credentials);
+    this.jobControllerClient = JobControllerClient.create(
+      JobControllerSettings.newBuilder().setCredentialsProvider(credentialsProvider)
+        .setEndpoint(region + DATAPROC_GOOGLEAPIS).build());
+  }
+
+  @Override
+  public RuntimeJobId launch(RuntimeJobInfo runtimeJobInfo) throws Exception {
+    ProgramRunInfo runInfo = runtimeJobInfo.getProgramRunInfo();
+    LOG.info("Starting to launch program run {} with following configurations: cluster {}, project {}, region {}, " +
+               "bucket {}.", runInfo.getRun(), clusterName, projectId, region, bucket);
+
+    File tempDir = Files.createTempDirectory("dataproc.launcher").toFile();
```

Comment:
Add a TODO to actually use a fixed directory and caching the twill/application/artifact jars, since they won't change if it is the same app.

Reply:
Do you mean caching on local directory only right? what about gcs?

Reply:
Both. Caching on local disk saves the time to rebuild the jar. Caching on GCS saves the upload time.